### PR TITLE
cloudconfig image pr

### PR DESCRIPTION
**Pull request by naveen-armedia**
RESOURCE_PATH is a dynamic argument where the jar file and startup script is added.   It will have the path where we build jar artifact.. For example jar is inside a `target` directory, docker image can be built with `docker build -t  arkcase-cloudconfig  --build-arg  RESOURCE_PATH= 'target'  <DockerfilePath> `

### CODE REVIEW

**Review by fabricetriboix — COMMENTED** [review]
I thought we would build from source using a multi-stage Docker build. @JonHolman Could you please comment?

**Comment by JonHolman** [comment]
> I thought we would build from source using a multi-stage Docker build. @JonHolman Could you please comment?

**Comment by JonHolman** [comment]
> I thought we would build from source using a multi-stage Docker build. @JonHolman Could you please comment?

Sorry for the accidental close and re-open.  @david-oc-miller should we be building cloud config within the docker file or get the WAR file from another source?  I think I'm understanding @fabricetriboix 's question there.

I believe we should use multi-stage dockerfiles when appropriate, I think that would be when we're building the JAR/WAR within our Dockerfile.

File: Dockerfile [hunk 1]
```diff
@@ -1 +1,16 @@
-FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/hello-world:latest
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base_java8:latest
+WORKDIR /app
+RUN useradd --create-home --user-group arkcase \
+        && mkdir /app/tmp \
+        && chown -R arkcase:arkcase /app 
+        
+    
+USER arkcase
+COPY --chown=arkcase ./config-server-2021.03-SNAPSHOT.jar /app/config-server.jar
```
Comment (david-oc-miller) [inline Dockerfile]: where does the file `config-server-2021.03-SNAPSHOT.jar` come from?  I don't see it in the repository or in the pull request.

File: start.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+set -eu
+
+repodir=/app/data
+search_locations="file:$repodir,file:$repodir/labels,file:$repodir/ldap,file:$repodir/lookups,file:$repodir/rules"
```
Comment (david-oc-miller) [inline start.sh]: How do the `labels`, `ldap`, `lookups`, and `rules` folders get created?  Also, the config server uses more folders: `branding`, `pdf-stylesheets` and `spring`.

Comment (david-oc-miller) [reply start.sh]: The whole `repodir` folder has to be shared with the ArkCase container, but I don't see any volume mounts or volumes or whatnot.  If the files in this MR are incomplete, the MR should be marked `WIP` ("work in progress").  Or else, the missing files should be added to the MR.

**Comment by david-oc-miller** [comment]
The config server is a jar file from the `acm-config-server` repository.  The acm-config-server repository already has a .gitlab-ci.yaml file that participates in the overall ArkCase build chain, which creates the jar file and uploads it into both Nexus and fileshare.armedia.com.  So we don't need to build the jar file in this repository.  

The config server image/container needs to serve configuration files; the configuration files are from the .arkcase repository.  The .arkcase repository has many branches, many of which have their own special .gitlab-ci.yaml files, so as to produce config files for different ArkCase flavors: core, FOIA, privacy, EEOC, UW, BCGEU, PRC, PAHO, DHS CAPS.  

I believe we need corresponding images; one config server image for each .arkcase branch that has its own .gitlab-ci.yaml.  If you want to package the config files along with the config server in this one config server image, you need a corresponding number of config server images.  Or, perhaps you could have one config server image, and many configuration images, where each configuration image matches a .arkcase branch.

File: Dockerfile (hunk 1 above)
Comment (david-oc-miller) [reply Dockerfile]: Maybe in this Dockerfile you could either provide (through an environment variable) or specify in the Dockerfile, what version of the config server you want to build, then you could retrieve that version from Nexus?

Comment (naveen-armedia) [reply Dockerfile]: jar file location will be given at build time with `RESOURCE_PATH` arg. Also added cloudconfig version as an argument. So when building both jar location and cloudconfig version will given at build time.

File: start.sh (hunk 2 above)
Comment (naveen-armedia) [reply start.sh]: .arkcase extention repo which is shared with arkcase core will be cloned at `ark_extension_path` location path. And value for `ark_extension_path` is given as a environment variable to the POD. For example if .arkcase repo is clone at `/data` the value of ark_extension_path will be given as `/data/.arkcase`

File: Dockerfile [hunk 3]
```diff
@@ -1 +1,22 @@
-FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/hello-world:latest
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base_java8:latest
+LABEL ORG="Armedia LLC" \
+      VERSION="1.0" \
+      IMAGE_SOURCE=https://github.com/ArkCase/ark_cloudconfig \
+      MAINTAINER="Armedia LLC"
+ARG RESOURCE_PATH="target"
+ARG CLOUD_CONFIG_VERSION=2021.03
```
Comment (JonHolman) [inline Dockerfile]: please combine these into one Dockerfile command

File: Dockerfile [hunk 4]
```diff
@@ -1 +1,22 @@
...
+WORKDIR /app
+RUN useradd --create-home --user-group arkcase \
+        && mkdir /app/tmp \
+        && chown -R arkcase:arkcase /app 
+
+USER arkcase
+#COPY the application war files
+COPY --chown=arkcase ${RESOURCE_PATH}/config-server-${CLOUD_CONFIG_VERSION}-SNAPSHOT.jar /app/config-server.jar 
+COPY --chown=arkcase ${RESOURCE_PATH}/start.sh /app/start.sh
```
Comment (JonHolman) [inline Dockerfile:16]: please combine these lines

File: Dockerfile (hunk 3 above)
Comment (fabricetriboix) [reply Dockerfile]: What is the rationale to add a `CLOUD_CONFIG_VERSION` argument? I would rather have that hardcoded in the Dockerfile. When a new version of the cloud config server is available, we just need to update the Dockerfile with the new version, commit and push. Then the pipeline will take care of producing a new image.

File: Dockerfile (hunk 4 above)
Comment (fabricetriboix) [reply Dockerfile:16]: I think a wrapper script should be provided that would (1) fetch the config-server jar file and then (2) run the `docker build` command.

File: conf/application.yaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  level:
+    com.armedia.acm.configserver: debug
+    org.springframework.cloud.config: debug
+  # pattern:
+  #   file: "%d{yyyy-MM-dd HH:mm:ss,SSS} [%thread] %-5level %logger.%M - %msg%n"
+  jms.message.buffer.window: 1
+  #activemq url and secrets will be templatized with helm 
+  acm.activemq:
+    broker-url: ssl://${activemq-url}:61616
+    keystore: /app/conf/arkcase.ks
```
Comment (fabricetriboix) [inline conf/application.yaml]: No keystore/truststore should be specified because we will run all the container on Kubernetes with Istio.

**Review by fabricetriboix — CHANGES_REQUESTED** [review]
Please remove the jar file from git.

File: Dockerfile (hunk 4 above)
Comment (JonHolman) [reply Dockerfile:16]: That would work.  Should we store the username/password in SSM parameter store and fetch them from that wrapper script?

File: Dockerfile (hunk 3 above)
Comment (naveen-armedia) [reply Dockerfile]: @JonHolman tried to combine into single docker command but got error when testing. After doing some research, ARG command only allows one argument.
@fabricetriboix  when we had call with Jhon Shinnick it was mentioned that version for config server needs to be dynamic so  `CLOUD_CONFIG_VERSION ` argument is added.  And with this we don't need to change dockerfile again when we need build with different version of cloud config server.

File: conf/application.yaml (hunk 5 above)
Comment (naveen-armedia) [reply conf/application.yaml]: Yes, removed it.

File: Dockerfile (hunk 4 above)
Comment (fabricetriboix) [reply Dockerfile:16]: @JonHolman I think that's a good idea

**Review by fabricetriboix — DISMISSED** [review]

File: Dockerfile (hunk 3 above)
Comment (JonHolman) [reply Dockerfile]: please make CLOUD_CONFIG_VERSION=2021.03 an ENV per our meeting Thursday

File: Dockerfile (hunk 4 above)
Comment (JonHolman) [reply Dockerfile:16]: per our meeting with @ymuwakki we will now be using an s3 bucket instead of our sftp

Comment (JonHolman) [reply Dockerfile:16]: still need a script to fetch the artifacts though

File: Dockerfile [hunk 6]
```diff
@@ -1 +1,22 @@
...
+ARG RESOURCE_PATH="artifacts" 
+ENV CLOUD_CONFIG_VERSION="2021.03"
+WORKDIR /app
+RUN useradd --create-home --user-group arkcase \
+        && mkdir /app/tmp \
+        && chown -R arkcase:arkcase /app 
+
+USER arkcase
+#COPY the application war files
+COPY --chown=arkcase ${RESOURCE_PATH}/config-server-${CLOUD_CONFIG_VERSION}-SNAPSHOT.jar /app/config-server.jar
```
Comment (JonHolman) [inline Dockerfile:15]: please combine these two COPY lines into one COPY

**Review by JonHolman — CHANGES_REQUESTED** [review]

**Review by fabricetriboix — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (fabricetriboix) [reply Dockerfile:16]: @JonHolman I think I am missing something here... I thought we said to store the artifacts in the S3 bucket? Or is it only for third-party artifacts and we store our own artifacts on fileshare.armedia.com?

File: Dockerfile (hunk 6 above)
Comment (naveen-armedia) [reply Dockerfile:15]: @JonHolman  we cannot combine two COPY instructions  into one here. As in the first COPY we are copying to the file to the destination with different name `config-server.jar` .

**Review by fabricetriboix — APPROVED** [review]

**Review by JonHolman — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (JonHolman) [reply Dockerfile:16]: I thought we were storing all artifacts in that s3 bucket.  The script I'm talking about is get-artifacts.sh to retrieve the artifacts from s3 prior to docker build.

Comment (fabricetriboix) [reply Dockerfile:16]: @JonHolman I still don't understand your comment about sftp...

Comment (JonHolman) [reply Dockerfile:16]: @fabricetriboix shoot, sorry.  That was supposed to say "now", not "not".  I was confused until you said sftp, then I went back to see what I said.  We are NOT using sftp for anything.  Sorry.

Comment (fabricetriboix) [reply Dockerfile:16]: @JonHolman All right, that makes sense now!

File: start.sh (hunk 2 above)
Comment (JonHolman) [reply start.sh]: @david-oc-miller @naveen-armedia  can this conversation be resolved?

Comment (naveen-armedia) [reply start.sh]: Yes @JonHolman

**Pull request merged**